### ISSUE TITLE
cpplint: virtual and final are flagged as an error

### ISSUE THREAD

**Issue by mpercy** [google/styleguide#176]
At the time of writing, the inheritance/readability rule checks for virtual on the same declaration as final, and marks it as an error. However, that syntax is required for C++11 (at least on clang++) so this error is not legit.

**Comment by mpercy** [comment]
Offending rule at time of writing is at https://github.com/google/styleguide/blob/gh-pages/cpplint/cpplint.py#L5569

**Comment by tituswinters** [comment]
Uh, that's not true.

You can't tag a non-virtual method as override or final, but that's sort of the point: it has to be virtual in the parent before those do anything.

What are you trying to do, specifically?

**Comment by mpercy** [comment]
Hi @tituswinters, thanks for your response. I work on Apache Kudu and cpplint flags required syntax as an error, so I have disabled the "readability/inheritance" rule in Kudu's LINT build as a workaround for now. This is the line that is flagged as an error: https://github.com/apache/kudu/blob/master/src/kudu/integration-tests/mini_cluster_base.h#L48

```
  virtual void Shutdown() final {
    ShutdownNodes(ClusterNodes::ALL);
  }
```

This is the text of the cpplint error message:

```
/home/mpercy/src/kudu/src/kudu/integration-tests/mini_cluster_base.h:48:  "virtual" is redundant since function is already declared as "final"  [readability/inheritance] [4]
Total errors found: 1
```

**Comment by mpercy** [comment]
I think I see the confusion here. The problem is that I want to declare a method as final in the superclass. cpplint doesn't allow that.

**Comment by tituswinters** [comment]
So the question becomes: why are you doing that? There's very little reason
to do so, in our experience.

**Comment by mpercy** [comment]
The method is a part of the contract of the base class interface, for which it provides an implementation that is not intended to be overridden.

In this case, I have a a version of `Shutdown()` called `ShutdownNodes()` that takes an argument. I would rather specify a default argument but doing that on virtual methods is broken in C++ and rightly banned by the style guide. So instead I provide a convenience no-arg `Shutdown()` method in the base class that delegates to `ShutdownNodes(ALL)`, which is implemented by the subclasses. There is no reason for a subclass to override this `Shutdown()` method so I would like to declare it `final` to indicate that.

**Comment by tituswinters** [comment]
But it literally doesn't work to override that if it isn't virtual. Why
complicate things? Just don't mark it virtual in the first place,
everything works.

**Comment by mpercy** [comment]
Making it non-virtual does not prevent overriding it in a broken way:

```
$ cat test.cc
#include <iostream>
#include <memory>

class Base {
 public:
  void Shutdown() {
    std::cout << "Base" << std::endl;
  }
};

class Derived : public Base {
 public:
  void Shutdown() {
    std::cout << "Derived" << std::endl;
  }
};

int main() {
  std::unique_ptr<Base> base(new Derived());
  base->Shutdown();
  return 0;
}

$ g++ --std=c++11 test.cc -o test
$ ./test
Base
```

**Comment by tituswinters** [comment]
Sure, but as your sample execution shows: it doesn't _do_ anything from the derived version.  If you try to make all user errors into compilation errors, where does it end?  Do you also want to make misspellings of shutdown a compilation error?  Or alternate Capitalizations? (ShutDown vs. Shutdown?) 

We've dug into this (for years) - there's nothing good down that path.  You can't completely prevent that class of errors - don't try.

**Comment by vapier** [comment]
public cpplint development in this repository has been halted (see #837).  please migrate to https://github.com/cpplint/cpplint now.
